service/matching/physical_task_queue_manager.go

```diff
@@ -128,12 +134,108 @@ type (
+}
+
+func (cb *circularTaskBuffer) advance() {
+	cb.currentPos = (cb.currentPos + 1) % len(cb.buffer)
+	cb.buffer[cb.currentPos] = 0 // Reset the task count for the new interval
+}
+
+// returns the total number of tasks in the buffer
+func (cb *circularTaskBuffer) totalTasks() int {
+	totalTasks := 0
+	for _, count := range cb.buffer {
+		totalTasks += count
+	}
+	return totalTasks
+}
+
+type taskTracker struct {
+	mutex             sync.RWMutex
+	clock             clock.TimeSource
+	startTime         time.Time     // time when taskTracker was initialized
+	startIntervalTime time.Time     // the starting time of a window in the buffer
+	interval          time.Duration // the duration of each mini- in the buffer
+	totalIntervalSize int           // the number over which rate of tasks added/dispatched would be determined
+	tasksInInterval   *circularTaskBuffer
+}
+
+func newTaskTracker(timeSource clock.TimeSource, intervalSize int, totalIntervalSize int) *taskTracker {
+	return &taskTracker{
+		clock:             timeSource,
+		startTime:         timeSource.Now(),
+		startIntervalTime: timeSource.Now(),
+		interval:          time.Duration(intervalSize) * time.Second, // Todo: Shivam - replace with config value
+		totalIntervalSize: totalIntervalSize,
+		tasksInInterval:   newCircularTaskBuffer((totalIntervalSize / intervalSize) + 1),
+	}
+}
+
+// advanceAndResetTracker is a helper to advance the trackers position and clear out any expired intervals
+func (s *taskTracker) advanceAndResetTracker(elapsed time.Duration) {
+	// Calculate the number of intervals elapsed since the start interval time
+	intervalsElapsed := int(elapsed / s.interval)
+
+	if intervalsElapsed != 0 {
+		for i := 0; i < intervalsElapsed; i++ {
+			s.tasksInInterval.advance() // advancing our circular buffer's position until we land on the right interval
+		}
+		s.startIntervalTime = s.startIntervalTime.Add(time.Duration(intervalsElapsed) * s.interval)
+	}
+}
+
+// trackTasks is responsible for adding/removing tasks from the current time that falls in the appropriate interval
+func (s *taskTracker) trackTasks() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	currentTime := s.clock.Now()
+
+	// Calculate elapsed time from the latest start interval time
+	elapsed := currentTime.Sub(s.startIntervalTime)
+	s.advanceAndResetTracker(elapsed)
+	s.tasksInInterval.incrementTaskCount()
+}
+
+// rate is responsible for returning the rate of tasks added/dispatched in a given interval
+func (s *taskTracker) rate() float32 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	currentTime := s.clock.Now()
+
+	// Calculate elapsed time from the latest start interval time
+	elapsed := currentTime.Sub(s.startIntervalTime)
+	s.advanceAndResetTracker(elapsed)
+	totalTasks := s.tasksInInterval.totalTasks()
+
+	// time passed since start of the current window + totalIntervalSize in milliseconds
+	elapsedTime := currentTime.Sub(s.startIntervalTime).Milliseconds() + int64(s.totalIntervalSize*1000)
```

Comment:
We should not use the full window length if the pq has not been running for that long.
```suggestion
	elapsedTime := min(currentTime.Sub(s.startIntervalTime).Milliseconds() + int64(s.totalIntervalSize*1000), currentTime.Sub(s.startTime).Milliseconds())
```

Reply:
two things about this:

1. I thought the discussion with @dnr was about averaging out the value and I worry this might not do the job since I had this previously and it was deemed wrong. Consider a case where the pq has been running for 1 second and 100 tasks have been added. Earlier, and in the suggest approach, the expected rate would be 100. However, I was under the assumption that we should "average" this value out by emitting an expected value and should account for a time frame of 30 seconds. Was this not what we were going for?

2. I fear that the suggested line change will always result in `currentTime.Sub(s.startIntervalTime).Milliseconds()` being chosen. Since we have 5 second intervals, `s.startIntervalTime` denotes the starting time of each 5 second start and we will never get the 30 second period as we had discussed.

Reply:
This works, although I am afraid we might be going back to something which I had initially:

1. I thought the discussion with @dnr was about averaging out the rate even if the pq was run for a short span of time. Consider a case where the pq has been running for 1 second and 100 tasks have been added. Earlier, and in your suggested approach, the expected rate would be 100. However, I was under the assumption that we should "average" this value out by emitting an `expected` value and should account for a time frame of 30 seconds. Was this not what we were going for?